airbyte-integrations/connectors/source-chargebee/unit_tests/integration/test_addon.py

```diff
@@ -0,0 +1,198 @@
+        )
+        output = self._read(_config().with_start_date(self._start_date))
+        assert len(output.records) == 2
+
+    @HttpMocker()
+    def test_given_multiple_pages_of_records_read_and_returned(self, http_mocker: HttpMocker) -> None:
+        # Tests pagination
+        request = _a_request().with_any_query_params().build()
+        http_mocker.get(
+            request,
+            [_a_response().with_record(_a_record()).with_pagination().build(), _a_response().with_record(_a_record()).build()]
+        )
+        output = self._read(_config().with_start_date(self._start_date - timedelta(hours=8)))
+        http_mocker.assert_number_of_calls(request, 2)
+        assert len(output.records) == 2
+
+    @HttpMocker()
+    def test_given_http_status_400_when_read_then_stream_is_ignored(self, http_mocker: HttpMocker) -> None:
+        # Tests 400 status error handling
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            a_response_with_status(400)
+        )
+        output = self._read(_config().with_start_date(self._start_date), expecting_exception=True)
+        assert len(output.get_stream_statuses(f"{_STREAM_NAME}s")) == 0
+
+
+    @HttpMocker()
+    def test_given_http_status_401_when_the_stream_is_incomplete(self, http_mocker: HttpMocker) -> None:
+        # Test 401 status error handling
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            a_response_with_status(401),
+        )
+        output = self._read(_config().with_start_date(self._start_date), expecting_exception=True)
+        assert output.errors[-1].trace.error.failure_type == FailureType.system_error
+
+    @HttpMocker()
+    def test_given_rate_limited_when_read_then_retry_and_return_records(self, http_mocker: HttpMocker) -> None:
+        # Tests backoff/retry with rate limiting
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            [
+                a_response_with_status_and_header(429, {"Retry-After": "0.01"}),
+                _a_response().with_record(_a_record()).build(),
+            ],
+        )
+        output = self._read(_config().with_start_date(self._start_date))
+        assert len(output.records) == 1
+
+    @HttpMocker()
+    def test_given_http_status_500_once_before_200_when_read_then_retry_and_return_records(self, http_mocker: HttpMocker) -> None:
+        # Tests retry with 500 status
+        http_mocker.get(
+            _a_request().with_any_query_params().build(),
+            [a_response_with_status_and_header(500, {"Retry-After": "0.01"}), _a_response().with_record(_a_record()).build()],
+        )
+        output = self._read(_config().with_start_date(self._start_date))
+        assert len(output.records) == 1
+
+    @HttpMocker()
+    def test_given_http_status_500_on_availability_when_read_then_raise_system_error(self, http_mocker: HttpMocker) -> None:
```

Comment:
I find this name a bit confusing as `test_given_http_status_500_once_before_200_when_read_then_retry_and_return_records` also has a 500 on availability but does not raise a system_error

Reply:
Updated name to differentiate between the two tests.